pymeasure/instruments/lecroy/lecroyT3DSO1204.py

```diff
@@ -0,0 +1,1192 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2022 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import logging
+import re
+import sys
+import time
+from decimal import Decimal
+
+import numpy as np
+
+from pymeasure.instruments import Instrument
+from pymeasure.instruments.validators import strict_discrete_set, strict_range
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+def _sanitize_source(source):
+    """ Parse source string
+
+    :param source can be "cX", "ch X", "chan X", "channel X", "math" or "line", where X is
+    a single digit integer. The parser is case and white space insensitive.
+    :return: can be "C1", "C2", "C3", "C4", "MATH" or "LINE. """
+
+    match = re.match(r"\s*(C\s*\d|CH\s*\d|CHAN\s*\d|CHANNEL\s*\d|MATH|LINE)", source,
+                     re.IGNORECASE)
+    if match:
+        source = match.group(1).replace(" ", "").upper()
+        if source not in ["MATH", "LINE"]:
+            source = "C" + source[-1]
```

Comment:
That is very good advice. I improved the channel name sanitizing method in the commit 577f61d36e198a8063a0088e9c5f83f96427c8fa

Reply:
Still a messy regex, but good enough.

When doing regexes, I try to always use named groups where appropriate. 
On the one hand, the regex becomes more readable as the "purpose" of a capturing group is clarified in the regex
On the other hand you can use the names later when pulling stuff out of the group object, which makes the code much clearer.
E.g., you can test with `if match.group("number") is None...`, etc.

I came up with this, but haven't run the tests:
```
r"^\s*(?:(?P<name>C|CH|CHAN|CHANNEL)\s*(?P<number>\d))|(?P<name_only>MATH|LINE)\s*$"
```
You can then (I think) simplify the logic to "if a number is found, use C + number, else use name_only"

You can try this out if you want.

Reply:
@bilderbuchi Thank you for the tip. I did not know about named capture groups.
I slightly modified the regex you suggested and it is working fine: f64564563d50ab0950a21bfec1f0f162610b3315

Reply:
Great! One thing: The way the regex is written you have three scenarios: 
* The first non-capturing group matches, and you'll have `number` **and** `name` groups (otherwise: no match at all), or
* the second group matches and you get `name_only`
* or there is no match at all and the `if match` will already fail.

This means, afaict, that your result testing logic can be simplified to a single if as I outlined above. Please test to be sure.

Reply:
I think that in the current version of the code the number of if/else statements is the minimum one required to catch the three outcomes that you listed.